File: solutions/csharp/rotational-cipher/1/RotationalCipher.cs
public static class RotationalCipher
{
    public static string Rotate(string text, int shiftKey)
    {
        var shift = shiftKey % 26;

        var cipher = text.Select(character => char.IsLetter(character) ? RotateLetter(character, shift) : character);

        return string.Concat(cipher);
    }

    private static char RotateLetter(char character, int shiftKey)
    {
        var alphabetStart = char.IsUpper(character) ? 'A' : 'a';

        return (char)(alphabetStart + (character - alphabetStart + shiftKey) % 26);
    }
}

File: solutions/scala/rotational-cipher/1/src/main/scala/RotationalCipher.scala
object RotationalCipher {
  private val AlphabetSize = 26

  def rotate(text: String, shift: Int): String =
    text.map(rotate(_, shift))

  private def rotate(character: Char, shift: Int): Char =
    character match {
      case c if c >= 'a' && c <= 'z' => rotateFrom(c, 'a', shift)
      case c if c >= 'A' && c <= 'Z' => rotateFrom(c, 'A', shift)
      case c => c
    }

  private def rotateFrom(character: Char, base: Char, shift: Int): Char =
    val offset = (character - base + shift) % AlphabetSize

    (base + offset).toChar
}
